smartpower3/wifimanager.cpp

```diff
@@ -46,295 +56,208 @@ WiFiManager::WiFiManager(WiFiUDP &udp, WiFiClient &client)
 	this->client = client;
 }
 
-void WiFiManager::do_yes_no_selection(void (WiFiManager::*func)(), const char *confirmation_string, const char *approval_string, const char *denial_string)
-{
-	char buf, cmd;
-
-	Serial.printf("%s", confirmation_string);
-
-	while (true) {
-		if (Serial.available()) {
-			cmd = Serial.read();
-			switch (cmd) {
-				case SERIAL_CTRL_C:
-					return;
-				case SERIAL_ENTER:
-					switch (buf) {
-						case SERIAL_Y:
-							(this->* func) ();
-							Serial.println(approval_string);
-							return;
-						case SERIAL_N:
-							Serial.println(denial_string);
-							return;
-						default:
-							Serial.printf(MSG_YN_REDO_SELECTION_REQUEST);
-					}
-					break;
-				default:
-					Serial.printf("%c", cmd);
-					buf = cmd;
-			}
-		}
-	}
-}
-
 void WiFiManager::view_main_menu(void)
 {
-	int i;
-
 	Serial.printf("\n\r");
-	for(i = 0; i < WIFI_CMD_MENU_CNT; i++)
+	for(uint8_t i = 0; i < WIFI_CMD_MENU_CNT; i++)
 		Serial.printf("%s\n\r", &WIFI_CMD_MENU[i][0]);
 
 	Serial.printf("%s", MSG_CMD);
 }
 
-void WiFiManager::view_ap_list(int ap_list_cnt)
+void WiFiManager::view_ap_list(int16_t ap_list_cnt)
 {
-	char ap_name[64], i;
-
-	if (ap_list_cnt) {
+	// scanNetworks in ap_scanning can return negative value indicating failure
+	// or unfinished scan - see WiFiType.h
+	if (ap_list_cnt > 0) {
 		Serial.printf("[ Networks found(%d) ]\n\r", ap_list_cnt);
-		for (i = 0; i < ap_list_cnt; i++) {
-			memset(ap_name, 0x00, sizeof(ap_name));
-			WiFi.SSID(i).toCharArray(ap_name, sizeof(ap_name));
+		for (int16_t i = 0; i < ap_list_cnt; i++) {
 			Serial.printf("%d - %s (%d dbm), Encryption (%s)\n\r",
 							i,
-							&ap_name[0],
+							WiFi.SSID(i).c_str(),
 							WiFi.RSSI(i),
 							encryption_str(WiFi.encryptionType(i)));
 		}
-	}
-	else
+	} else if (!ap_list_cnt) {
 		Serial.printf("%s\n\r", MSG_CMD_NO_NETWORK);
+	} else if (ap_list_cnt == WIFI_SCAN_RUNNING) {
+		Serial.printf("%s\n\r", MSG_CMD_SCAN_RUNNING);
+	} else if (ap_list_cnt == WIFI_SCAN_FAILED) {
+		Serial.printf("%s\n\r", MSG_CMD_SCAN_FAILED);
+	} else {
+		Serial.printf("%s\n\r", MSG_CMD_NETWORK_UNKNOWN_ERROR);
+	}
 }
 
-void WiFiManager::ap_scanning(void)
+int16_t WiFiManager::ap_scanning(void)
 {
 	Serial.printf("%s\n\r", MSG_CMD_SCANNING);
-	APListCount = WiFi.scanNetworks();
+	WiFi.disconnect();  // needs to be disconnected between successive scans
+	return WiFi.scanNetworks();
```

Comment:
WiFi needs to be disconnected between successive scans, otherwise the scan can return negative values indicating error, scan in progress etc. When connecting to AP after the scan, everything was OK. But if the connection was not made (for example user enter wrong password) and the scan repeated, that led to "scan in progress" being returned.

Fixes https://github.com/hardkernel/smartpower3/issues/20